agent/agent.go

```diff
@@ -3202,6 +3208,83 @@ func (a *Agent) loadProxies(conf *config.RuntimeConfig) error {
 	return persistenceErr
 }
 
+func (a *Agent) getPersistedTokens() (map[string]string, error) {
+	if !a.config.ACLEnableTokenPersistence {
+		return nil, nil
+	}
+
+	a.persistedTokensLock.RLock()
+	defer a.persistedTokensLock.RUnlock()
+
+	tokensFullPath := filepath.Join(a.config.DataDir, tokensPath)
+
+	buf, err := ioutil.ReadFile(tokensFullPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// non-existence is not an error we care about
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed reading tokens file %q: %s", tokensFullPath, err)
+	}
+
+	persistedTokens := make(map[string]string)
```

Comment:
What's the reason for having this be a map[string]string instead of a typed struct (something like config.Tokens)?

Reply:
The main reason was that it was easier to check if a value existed or not in addition to being the empty string if a regular map was used. If we decode into a struct much of that information is lost. In this case we need to know whether its okay to pull a token from a config. We prefer persisted tokens to tokens within the configuration and emit warnings if both are used. However its valid and somewhat expected to use both ways simultaneously. In particular I would expect it to be common to set the agent_master token via the config and all the rest via the API.

Reply:
It looks like everywhere that checks whether a token is set also checks for a non-empty-string, so it doesn't seem like we actually use the info on whether the values are set (so using a struct would be the same, unless I've missed somewhere).

Reply:
Yeah you are right. From my original design I said "empty string values will be treated like the same as the map not containing the key at all."

So yes we could decode into a struct.